mutation/mutation.cc

```diff
@@ -194,6 +196,89 @@ mutation reverse(mutation mut) {
     return *std::move(mut).consume(reverse_rebuilder, consume_in_reverse::yes).result;
 }
 
+namespace {
+class mutation_by_size_splitter {
+    struct partition_state {
+        mutation_rebuilder_v2 builder;
+        size_t empty_partition_size;
+        size_t size = 0;
+        explicit partition_state(schema_ptr schema)
+            : builder(std::move(schema))
+        {
+        }
+    };
+    const schema_ptr _schema;
+    std::vector<mutation>& _target;
+    const size_t _max_size;
+    std::optional<partition_state> _state;
+    template <typename T>
+    stop_iteration consume_fragment(T&& fragment) {
+        const auto fragment_size = fragment.memory_usage(*_schema);
+        if (_state->size && _state->size + _state->empty_partition_size + fragment_size > _max_size) {
+            _target.emplace_back(_state->builder.flush());
+            // We could end up with an empty mutation if we consumed a range_tombstone_change
+            // and the next fragment exceeds the limit. The tombstone range may not have been
+            // closed yet and range_tombstone will not be created.
+            // This should be a rare case though, so just pop such mutation.
+            if (_target.back().partition().empty()) {
+                _target.pop_back();
+            }
+            _state->size = 0;
+        }
+        _state->size += fragment_size;
+        _state->builder.consume(std::move(fragment));
+        return stop_iteration::no;
+    }
+public:
+    mutation_by_size_splitter(schema_ptr schema, std::vector<mutation>& target, size_t max_size)
+        : _schema(std::move(schema))
+        , _target(target)
+        , _max_size(max_size)
+    {
+    }
+    void consume_new_partition(const dht::decorated_key& dk) {
+        _state.emplace(_schema);
+        _state->empty_partition_size = _state->builder.consume_new_partition(dk).memory_usage(*_schema);
+    }
+    void consume(tombstone t) {
+        _state->builder.consume(t);
+    }
+    stop_iteration consume(static_row&& sr) {
+        return consume_fragment(std::move(sr));
+    }
+    stop_iteration consume(clustering_row&& cr) {
+        return consume_fragment(std::move(cr));
+    }
+    stop_iteration consume(range_tombstone_change&& rtc) {
+        return consume_fragment(std::move(rtc));
+    }
+    stop_iteration consume_end_of_partition() {
+        _state->builder.consume_end_of_partition();
+        auto mut_opt = _state->builder.consume_end_of_stream();
+        assert(mut_opt);
```

Comment:
Please avoid using asserts, they can bring down entire clusters. Use `on_internal_error()` instead.

Reply:
> they can bring down entire clusters

can you elaborate?

Reply:
We compile `assert()` into release builds and an `assert()` triggering will cause ScyllaDB to crash.
It is bad enough if this happens on a single node, but there has been cases where multiple nodes or entire clusters were brought down by `asserts()` triggering.
So asserts should be avoided as much as possible, and always prefer to use `on_internal_error()` instead.
Use `assert()` only to check invariants, the violation of which is unrecoverable and the node is not able to continue working correctly.

Reply:
but in this case there are corresponding `assert`s inside `builder`, and we can quickly infer that if those assertions pass then this assertion has to pass too.
(Need to double-check)

Reply:
Double checked, actually there's no such assert. (We're using `mutation_rebuilder_v2`, not `mutation_rebuilder` which has it)

The logic is correct I think, we only call `end_of_partition()` when we have an opened partition.

But I agree  `on_internal_error` would be better.